File: apps/ios/Loonext/Features/Settings/VoiceGreeting.swift
import AVFoundation
import Foundation

/**
 #309 — a greeting recorded in the owner's own voice.

 A two-person outfit competing with a franchise sells on being a real, local,
 reachable person. Then nobody answers and the product hands the caller a
 synthetic voice reading a company name, which in 2026 is what a spam call
 sounds like.

 The recorder below is deliberately close to `WrapUpDictationRecorder`, because
 that one has already been through the audio-session edge cases. It differs in
 one way that matters: a dictation is heard once by a transcriber, and a
 greeting is heard by every customer who calls — so the sample rate is set for
 LISTENING rather than for speech recognition.
 */
struct VoicemailGreeting: Codable, Sendable, Identifiable {
    var id = ""
    var name = ""
    var duration_ms = 0
    var mime_type = ""
    var byte_size = 0
    var created_at = ""
}

/// Two minutes, the same ceiling the API and the column enforce.
let maxGreetingMs = 120_000

/// "0:08" — a duration a person reads, not 8200.
func formatGreetingDuration(_ ms: Int) -> String {
    let total = max(0, ms / 1000)
    return "\(total / 60):\(String(format: "%02d", total % 60))"
}

/// Why a recording could not start, in words the owner can act on.
enum GreetingStartRefusal {
    case callInProgress
    case noPermission
    case couldNotStart

    var message: String {
        switch self {
        case .callInProgress:
            return "You are on a call. End it and try again."
        case .noPermission:
            return "Loonext needs the microphone to record a greeting. "
                + "Allow it in iOS Settings, then try again."
        case .couldNotStart:
            return "The microphone is not available right now. Try again."
        }
    }
}

/// One recorded take, and the file it lives in until it is saved or thrown away.
struct GreetingTake {
    let audio: Data
    let durationMs: Int
    let url: URL
}

/**
 Owns what SwiftUI cannot: the live `AVAudioRecorder`, the file underneath it,
 and the player that hears it back.
 */
@MainActor
final class GreetingRecorder: ObservableObject {
    private var recorder: AVAudioRecorder?
    private var player: AVAudioPlayer?
    private var fileURL: URL?
    private var startedAt = Date.distantPast
    private var sessionActive = false

    var isRecording: Bool { recorder != nil }

    /// Delete any greeting audio a previous run left behind.
    ///
    /// Every in-process path deletes its own file, but none of them runs if the
    /// app is killed mid-record. Named by prefix so it can only ever remove
    /// this feature's own files.
    static func sweepOrphans() {
        let fm = FileManager.default
        guard let entries = try? fm.contentsOfDirectory(
            at: fm.temporaryDirectory,
            includingPropertiesForKeys: nil
        ) else { return }
        for url in entries where url.lastPathComponent.hasPrefix("greeting-") {
            try? fm.removeItem(at: url)
        }
    }

    func start(callInProgress: Bool) -> GreetingStartRefusal? {
        if isRecording { return nil }
        if callInProgress { return .callInProgress }

        // `.record`, never `.playAndRecord`: this app treats a `.playAndRecord`
        // session as "a call owns the audio", so borrowing that category here
        // would make an attachment play out of the earpiece afterwards. The
        // playback below re-activates its own session for the same reason.
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.record, mode: .spokenAudio)
            try session.setActive(true)
            sessionActive = true
        } catch {
            return .couldNotStart
        }

        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("greeting-\(UUID().uuidString).m4a")
        // Mono 44.1 kHz AAC. A customer hears this and the owner hears it back
        // before saving, so it is sized for LISTENING — where the dictation
        // recorder next door uses 16 kHz, which is right for a transcript and
        // thin for a greeting. Two minutes still lands well inside 2 MB.
        let settings: [String: Any] = [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 44100.0,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue,
        ]
        let made: AVAudioRecorder
        do {
            made = try AVAudioRecorder(url: url, settings: settings)
        } catch {
            deactivate()
            return .couldNotStart
        }
        // The platform's own stop, under the UI's timer rather than instead of
        // it: a view that misses its tick must not leave a microphone open.
        guard made.record(forDuration: TimeInterval(maxGreetingMs) / 1000) else {
            deactivate()
            try? FileManager.default.removeItem(at: url)
            return .couldNotStart
        }
        recorder = made
        fileURL = url
        startedAt = Date()
        return nil
    }

    /// Stop and hand back the take. Nil means nothing worth keeping was caught.
    func finish() -> GreetingTake? {
        guard let made = recorder, let url = fileURL else {
            discard()
            return nil
        }
        made.stop()
        recorder = nil
        deactivate()
        guard let data = try? Data(contentsOf: url), !data.isEmpty else {
            try? FileManager.default.removeItem(at: url)
            fileURL = nil
            return nil
        }
        return GreetingTake(
            audio: data,
            durationMs: Int(Date().timeIntervalSince(startedAt) * 1000),
            url: url
        )
    }

    /**
     Hear it back.

     Not a nicety: the whole flow rests on an owner hearing their own take
     before it becomes what customers hear. Without this the card would ask
     them to trust a recording they have never heard.
     */
    func play(_ take: GreetingTake) -> Bool {
        player?.stop()
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.playback, mode: .spokenAudio)
        try? session.setActive(true)
        guard let made = try? AVAudioPlayer(contentsOf: take.url) else { return false }
        player = made
        return made.play()
    }

    /// Throw the take away, and the bytes with it.
    func discard() {
        recorder?.stop()
        recorder = nil
        player?.stop()
        player = nil
        if let url = fileURL {
            try? FileManager.default.removeItem(at: url)
        }
        fileURL = nil
        deactivate()
    }

    private func deactivate() {
        guard sessionActive else { return }
        try? AVAudioSession.sharedInstance().setActive(false)
        sessionActive = false
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/VoiceGreeting.kt
package com.loonext.android.features.settings

import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import com.loonext.android.core.auth.await
import com.loonext.android.core.i18n.AppStrings
import com.loonext.android.core.net.ApiClient
import com.loonext.android.core.net.ApiErrorCode
import com.loonext.android.core.net.ApiException
import java.io.File
import java.io.IOException
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * #309 — a greeting recorded in the owner's own voice.
 *
 * A two-person outfit competing with a franchise sells on being a real, local,
 * reachable person. Then nobody answers and the product hands the caller a
 * synthetic voice reading a company name, which in 2026 is what a spam call
 * sounds like.
 *
 * The recorder below is deliberately close to [com.loonext.android.features
 * .compose.WrapUpDictation]'s, because that one has already been through the
 * OEM edge cases. It differs in one way that matters: a dictation is heard once
 * by a transcriber, and a greeting is heard by every customer who calls. So the
 * bitrate is set for LISTENING rather than for intelligibility — still under a
 * megabyte for two minutes, which the 2 MB ceiling covers comfortably.
 */
@Serializable
data class VoicemailGreeting(
    val id: String,
    val name: String,
    val duration_ms: Int,
    val mime_type: String = "",
    val byte_size: Int = 0,
    val created_at: String = "",
)

/** AAC in an MPEG-4 container: what [GreetingRecorder] writes. */
private const val GREETING_CONTENT_TYPE = "audio/mp4"
private const val GREETING_FILE_NAME = "greeting.m4a"

/** Two minutes, the same ceiling the API and the column enforce. */
const val MAX_GREETING_MS = 120_000

/** "0:08" — a duration a person reads, not 8200. */
fun formatGreetingDuration(ms: Int): String {
    val total = (ms / 1000).coerceAtLeast(0)
    return "${total / 60}:${(total % 60).toString().padStart(2, '0')}"
}

/**
 * The multipart door for a greeting.
 *
 * Borrows [ApiClient]'s OkHttp client, its fresh-session refresh and its
 * envelope decoding, exactly as the wrap-up transcriber next door does and for
 * the same reason: ApiClient speaks JSON bodies and keeps its base URL private.
 */
class GreetingUploader(private val api: ApiClient, private val baseUrl: String) {

    /**
     * [locale] is the READER's, carried in because this is not a composable and
     * the two refusals below are ours rather than the server's. Null asks for
     * the English table, which is what every caller had before it existed.
     */
    suspend fun upload(
        companyId: String,
        name: String,
        durationMs: Int,
        audio: ByteArray,
        locale: String? = null,
    ): VoicemailGreeting {
        val session = api.freshSession() ?: throw ApiException(
            ApiErrorCode.UNAUTHORIZED,
            AppStrings.translate(locale, "settingsMore.signedOut"),
            401,
        )
        val body = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart("name", name)
            .addFormDataPart("duration_ms", durationMs.toString())
            .addFormDataPart(
                "file",
                GREETING_FILE_NAME,
                audio.toRequestBody(GREETING_CONTENT_TYPE.toMediaTypeOrNull()),
            )
            .build()
        val request = Request.Builder()
            .url("$baseUrl/v1/voicemail-greetings")
            .header("Authorization", "Bearer ${session.accessToken}")
            .header("X-Company-Id", companyId)
            .post(body)
            .build()
        val response = try {
            api.http.newCall(request).await()
        } catch (_: IOException) {
            throw ApiException(
                ApiErrorCode.NETWORK,
                AppStrings.translate(locale, "settingsMore.cantReachLoonext"),
                0,
            )
        }
        val (status, text) = response.use { it.code to it.body?.string().orEmpty() }
        val payload = ApiClient.RawResponse(status, text).expectSuccess(api.json)
        return api.json.decodeFromString(payload)
    }
}

/** What [GreetingRecorder.finish] hands back. */
data class GreetingTake(val bytes: ByteArray, val durationMs: Int, val file: File) {
    // Kotlin warns on a data class holding a ByteArray, and it is right: two
    // takes with identical audio are not the same take. Identity is the file.
    override fun equals(other: Any?) = this === other
    override fun hashCode() = System.identityHashCode(this)
}

/**
 * Records one take to the cache directory.
 *
 * cacheDir, deliberately: until the owner saves it this is the most disposable
 * thing on the device, and if the process dies mid-record the system reclaiming
 * it is the correct outcome for audio nobody promised to keep.
 */
class GreetingRecorder(private val context: Context) {
    private var recorder: MediaRecorder? = null
    private var file: File? = null
    private var startedAtMs = 0L

    val isRecording: Boolean get() = recorder != null

    /**
     * Begin. Returns false when the platform refused — a microphone held by a
     * call, an OEM that rejects the encoder, a permission revoked between the
     * check and here. Nothing is left running on a false.
     */
    fun start(): Boolean {
        if (recorder != null) return true
        val target = try {
            File.createTempFile("greeting-", ".m4a", context.cacheDir)
        } catch (_: IOException) {
            return false
        }
        // Published BEFORE arming, so a discard() racing prepare()/start() on
        // the main thread finds the file and can honour its own cancellation.
        file = target
        val next = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            MediaRecorder(context)
        } else {
            @Suppress("DEPRECATION")
            MediaRecorder()
        }
        return try {
            next.setAudioSource(MediaRecorder.AudioSource.MIC)
            next.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            next.setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            // Sized for LISTENING, not for a transcript: a customer hears this,
            // and the owner hears it back before saving. Mono 44.1 kHz at
            // 64 kbps puts two minutes near 950 KB, inside the 2 MB ceiling.
            next.setAudioChannels(1)
            next.setAudioSamplingRate(44_100)
            next.setAudioEncodingBitRate(64_000)
            // The platform's own stop, under the UI's timer rather than instead
            // of it: a composition that misses its tick must not leave a
            // microphone open, and neither cap may be the only one.
            next.setMaxDuration(MAX_GREETING_MS)
            next.setOutputFile(target.absolutePath)
            next.prepare()
            next.start()
            if (file !== target) {
                runCatching { next.release() }
                target.delete()
                return false
            }
            recorder = next
            startedAtMs = System.currentTimeMillis()
            true
        } catch (_: Exception) {
            runCatching { next.release() }
            target.delete()
            if (file === target) file = null
            false
        }
    }

    fun elapsedMs(): Int =
        if (recorder == null) 0 else (System.currentTimeMillis() - startedAtMs).toInt()

    /** Stop and read the take back. Null when nothing usable was captured. */
    fun finish(): GreetingTake? {
        val active = recorder ?: return null
        val target = file
        recorder = null
        val elapsed = (System.currentTimeMillis() - startedAtMs).toInt()
        runCatching { active.stop() }
        runCatching { active.release() }
        if (target == null || !target.exists() || target.length() == 0L) {
            target?.delete()
            file = null
            return null
        }
        return GreetingTake(target.readBytes(), elapsed, target)
    }

    /** Throw the take away, and the bytes with it. */
    fun discard() {
        val active = recorder
        recorder = null
        if (active != null) {
            runCatching { active.stop() }
            runCatching { active.release() }
        }
        file?.delete()
        file = null
    }
}
